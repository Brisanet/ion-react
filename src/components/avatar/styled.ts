import { SizeType } from '@ion/core/types/size';
import styled, { RuleSet, css } from 'styled-components';

type AvatarProps = {
  $size: SizeType;
};

type initialsFontSizes = 12 | 14 | 16;

const sizes: (size: SizeType) => RuleSet<object> = (size) => {
  const sizeOptions = {
    lg: 40,
    md: 32,
    sm: 24,
    xs: 16,
  };

  return css`
    width: ${sizeOptions[size]}px;
    height: ${sizeOptions[size]}px;
  `;
};

const fontSize: (size: SizeType) => initialsFontSizes = (size) => {
  const sizeOptions = {
    lg: 16,
    md: 14,
    sm: 12,
    xs: 12,
  } as const;

  return sizeOptions[size];
};

export const AvatarContainer = styled.div<AvatarProps>`
  ${({ theme, $size }) => css`
    ${sizes($size)}
    border-radius: 50%;
    background-color: ${theme.colors.primary[2]};
    overflow: hidden;
    ${theme.utils.flex.center()}

    span {
      color: ${theme.colors.primary[7]};
      ${theme.font.size[fontSize($size)]};
      font-weight: 600;
    }
  `}
`;

export const AvatarPhoto = styled.img`
  object-fit: cover;
  height: 100%;
  width: 100%;
`;
