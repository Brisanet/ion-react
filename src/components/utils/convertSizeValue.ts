/**
 * Get a size value based on its type.
 * @param {string | number} value - The size value.
 * @returns The value itself if it's a string, or the value in pixels if it is a number.
 **/
export const convertSizeValue = (value: string | number): string =>
  typeof value === 'number' ? `${value}px` : value;
