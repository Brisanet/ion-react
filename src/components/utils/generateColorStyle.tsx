// Tipo para objeto de estilos  da função geradora de estilos
type ColorStyle = {
    background: string;
    color: string;
    border?: string;
  };
  
  /* 
    Função que gera cores de Background, Border e cor de
    fonte a fim de reduzir repetições de codigo
  
    Seguir o padrão de parâmetros: 
    1- Cor do background;
    2- Cor da fonte do botão;
    3- (Opicional) Cor da borda;
  */
 export function generateColorStyle(
    bgColor: string,
    fontColor: string,
    borderColor: string
  ) {
    let styles: ColorStyle = {
      background: bgColor,
      color: fontColor,
    };
  
    borderColor && (styles.border = `1px solid ${borderColor}`);
  
    return styles;
  }