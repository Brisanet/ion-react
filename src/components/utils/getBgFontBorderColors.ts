interface ObjStyles {
  background: string;
  color: string;
  border?: string;
}

/**
 * Function to prevent repetition of style blocks.
 *
 * @param background - Background color
 * @param color - The font color
 * @param border - Border attribute styles - example: ´1px solid blue´
 * @returns A Javascript object containing color properties.
 */
export function getBgFontBorderColors({
  background,
  color,
  border = '',
}: ObjStyles): ObjStyles {
  return {
    background,
    color,
    border,
  };
}
