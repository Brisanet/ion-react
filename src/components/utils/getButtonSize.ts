/**
 * Function to prevent repetition of style blocks.
 *
 * @param buttonSize - Height and min-width
 * @param padding - Button padding
 * @param fontSize - The size of font
 * @returns An Javascript object containing size properties.
 */
export function getButtonSize(
  buttonSize: string,
  padding: string,
  fontSize: string
) {
  let size = {
    height: buttonSize,
    minWidth: buttonSize,
    padding: padding,
    fontSize: fontSize,
  };

  return size;
}
