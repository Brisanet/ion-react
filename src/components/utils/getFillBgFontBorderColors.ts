type ColorStyleResponse = {
  border: string;
  backgroundColor: string;
  color: string;
  svg: {
    fill: string;
  };
};

interface StyleProps {
  svgFillColor: string;
  backgroundColor: string;
  fontColor: string;
  borderColor: string;
}

/**
 * Function to prevent repetition of style blocks.
 *
 * @param svgFillColor - Fill color to be used in SVG
 * @param backgroundColor - Background color
 * @param fontColor - The font color
 * @param borderColor - Border color
 * @returns A Javascript object containing style properties.
 */
export function getFillBgFontBorderColors({
  backgroundColor,
  fontColor,
  borderColor,
  svgFillColor,
}: StyleProps) {
  let styles: ColorStyleResponse = {
    backgroundColor: backgroundColor,
    color: fontColor,
    border: `1px solid ${borderColor}`,
    svg: {
      fill: svgFillColor,
    },
  };

  return styles;
}
