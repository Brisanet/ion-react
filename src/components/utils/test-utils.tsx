import theme from '@ion/styles/theme';
import { render, RenderResult } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';

export const renderWithTheme = (children: React.ReactNode): RenderResult =>
  render(<ThemeProvider theme={theme}>{children}</ThemeProvider>);
