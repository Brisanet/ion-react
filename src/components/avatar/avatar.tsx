import { SizeType } from '@ion/core/types/size';
import { AvatarContainer, AvatarPhoto } from './styled';
import { IonIcon } from '../icons';
import { useEffect, useState } from 'react';

export type AvatarType = 'initials' | 'icon' | 'photo';

export type AvatarProps = {
  type?: AvatarType;
  size?: SizeType;
  value?: string;
  image?: string;
  onErrorImg?: string;
};

const getInitials = (name: string): string =>
  name
    .split(' ')
    .map((word) => word[0])
    .slice(0, 2)
    .join('')
    .toUpperCase();

const defaultSize = 'md';
const defaultType = 'initials';

export const IonAvatar = ({
  type = defaultType,
  size = defaultSize,
  value,
  image,
  onErrorImg,
}: AvatarProps) => {
  const [errorImage, setErrorImage] = useState<string | null>(null);

  const initials = value ? getInitials(value) : '--';

  const iconSize = {
    lg: 24,
    md: 20,
    sm: 16,
    xs: 12,
  }[size];

  useEffect(() => {
    setErrorImage(null);
  }, [image]);

  const avatarTypes = {
    initials: <span data-testid='ion-avatar-initials'>{initials}</span>,
    icon: <IonIcon type='user' color='#06439d' size={iconSize} />,
    photo: image && (
      <AvatarPhoto
        data-testid='ion-avatar-photo'
        src={errorImage || image}
        onError={() => onErrorImg && setErrorImage(onErrorImg)}
      />
    ),
  };

  return (
    <AvatarContainer data-testid='ion-avatar' $size={size}>
      {avatarTypes[type]}
    </AvatarContainer>
  );
};
