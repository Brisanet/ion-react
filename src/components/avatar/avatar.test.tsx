import { fireEvent, screen } from '@testing-library/dom';
import { renderWithTheme } from '../utils/test-utils';
import { AvatarProps, AvatarType, IonAvatar } from './avatar';
import { SizeType } from '@ion/core/types/size';

const photoUrl =
  'https://64.media.tumblr.com/40e2174ab5e68b1eabbc3dfc78607cef/c1effc67d5c3a1fd-20/s540x810/9d6ce72fcddf97841e7410a0652dd9d5f018b35d.pnj';

const onErrorImg =
  'https://images.unsplash.com/photo-1714659046842-7dcf4ab2a58f?q=80&w=1974&auto=format&fit=crop&ixlib=rb-4.0.3&ixid=M3wxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8fA%3D%3D';

const avatarSizes: SizeType[] = ['lg', 'md', 'sm', 'xs'];

const photoAvatar: AvatarProps = {
  type: AvatarType.photo,
  image: photoUrl,
};

const sut = (props: AvatarProps) => renderWithTheme(<IonAvatar {...props} />);

describe('Avatar', () => {
  describe('Default', () => {
    it('should render avatar component', () => {
      sut({});
      expect(screen.getByTestId('ion-avatar')).toBeVisible();
    });
    it('should render the initials avatar as default', () => {
      sut({});
      expect(screen.getByTestId('ion-avatar-initials')).toBeVisible();
    });
    it.each(avatarSizes)(
      'should render avatar with %s size variation',
      (size) => {
        sut({
          size,
        });
        expect(screen.getByTestId('ion-avatar')).toMatchSnapshot();
      }
    );
  });

  describe('Initials', () => {
    it('should render with -- when no value is informed', () => {
      sut({});
      expect(screen.getByText('--')).toBeVisible();
    });
    it('should render the initials of a provided name', () => {
      sut({ value: 'ana carolina' });
      expect(screen.getByText('AC')).toBeVisible();
    });
  });

  describe('Icon', () => {
    it('should render a icon', () => {
      sut({ type: AvatarType.icon });
      expect(screen.getByTestId('ion-icon-user')).toBeVisible();
    });
  });

  describe('Photo', () => {
    it('should render the image', () => {
      sut(photoAvatar);
      expect(screen.getByTestId('ion-avatar-photo')).toBeVisible();
    });

    it('should change the image to the specified default on error', () => {
      sut({
        ...photoAvatar,
        image: 'wrongImageUrl',
        onErrorImg,
      });
      const imageElement = screen.getByTestId('ion-avatar-photo');
      fireEvent.error(imageElement);

      expect(imageElement).toHaveAttribute('src', onErrorImg);
    });
  });
});
