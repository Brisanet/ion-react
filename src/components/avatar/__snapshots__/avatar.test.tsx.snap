// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar Default should render avatar with lg size variation 1`] = `
.c0 {
  width: 40px;
  height: 40px;
  border-radius: 50%;
  background-color: #b5d2fc;
  overflow: hidden;
  display: flex;
  align-items: center;
  justify-content: center;
  gap: 0px;
}

.c0 span {
  color: #06439d;
  font-size: 1.6rem;
  line-height: 2.4rem;
  font-weight: 600;
}

<div
  class="c0"
  data-testid="ion-avatar"
>
  <span
    data-testid="ion-avatar-initials"
  >
    --
  </span>
</div>
`;

exports[`Avatar Default should render avatar with md size variation 1`] = `
.c0 {
  width: 32px;
  height: 32px;
  border-radius: 50%;
  background-color: #b5d2fc;
  overflow: hidden;
  display: flex;
  align-items: center;
  justify-content: center;
  gap: 0px;
}

.c0 span {
  color: #06439d;
  font-size: 1.4rem;
  line-height: 2rem;
  font-weight: 600;
}

<div
  class="c0"
  data-testid="ion-avatar"
>
  <span
    data-testid="ion-avatar-initials"
  >
    --
  </span>
</div>
`;

exports[`Avatar Default should render avatar with sm size variation 1`] = `
.c0 {
  width: 24px;
  height: 24px;
  border-radius: 50%;
  background-color: #b5d2fc;
  overflow: hidden;
  display: flex;
  align-items: center;
  justify-content: center;
  gap: 0px;
}

.c0 span {
  color: #06439d;
  font-size: 1.2rem;
  line-height: 1.6rem;
  font-weight: 600;
}

<div
  class="c0"
  data-testid="ion-avatar"
>
  <span
    data-testid="ion-avatar-initials"
  >
    --
  </span>
</div>
`;

exports[`Avatar Default should render avatar with xs size variation 1`] = `
.c0 {
  width: 16px;
  height: 16px;
  border-radius: 50%;
  background-color: #b5d2fc;
  overflow: hidden;
  display: flex;
  align-items: center;
  justify-content: center;
  gap: 0px;
}

.c0 span {
  color: #06439d;
  font-size: 1.2rem;
  line-height: 1.6rem;
  font-weight: 600;
}

<div
  class="c0"
  data-testid="ion-avatar"
>
  <span
    data-testid="ion-avatar-initials"
  >
    --
  </span>
</div>
`;
